provisioning/plugins/corim-psa-decoder/extractor.go

```diff
@@ -26,61 +43,61 @@ func (o Extractor) SwCompExtractor(rv comid.ReferenceValue) ([]*proto.Endorsemen
 	// reference-triple-record SHALL completely describe the updatable PSA RoT.
-	swComponents := make([]*proto.Endorsement, 0, len(rv.Measurements))
-
+	refVals := make([]*proto.Endorsement, 0, len(rv.Measurements))
+	var refVal *proto.Endorsement
+	var err error
 	for i, m := range rv.Measurements {
-		var psaSwCompAttrs PSASwCompAttributes
-
-		if err := psaSwCompAttrs.FromMeasurement(m); err != nil {
-			return nil, fmt.Errorf("extracting measurement at index %d: %w", i, err)
+		if m.Key == nil {
+			return nil, fmt.Errorf("measurement key is not present")
 		}
-
-		swAttrs, err := makeSwAttrs(psaClassAttrs, psaSwCompAttrs)
-		if err != nil {
-			return nil, fmt.Errorf("failed to create software component attributes: %w", err)
+		if !m.Key.IsSet() {
+			return nil, fmt.Errorf("measurement key is not set")
 		}
-
-		swComponent := proto.Endorsement{
-			Scheme:     proto.AttestationFormat_PSA_IOT,
-			Type:       proto.EndorsementType_REFERENCE_VALUE,
-			Attributes: swAttrs,
+		// Check which MKey is present and then decide which extractor to invoke
+		if m.Key.IsPSARefValID() {
+			var psaSwCompAttrs PSASwCompAttributes
+
+			refVal, err = ExtractMeas(&psaSwCompAttrs, m, psaClassAttrs)
+			if err != nil {
+				return nil, fmt.Errorf("unable to extract measurement at index %d, %w", i, err)
+			}
 		}
-
-		swComponents = append(swComponents, &swComponent)
+		if m.Key.IsCCAPlatformConfigID() {
+			if o.Profile != ccaProfile {
+				return nil, fmt.Errorf("measurement error at index %d: incorrect profile %s", i, o.Profile)
+			}
+			var ccaPlatformConfigID CCAPlatformConfigID
+			refVal, err = ExtractMeas(&ccaPlatformConfigID, m, psaClassAttrs)
+			if err != nil {
+				return nil, fmt.Errorf("unable to extract measurement: %w", err)
+			}
+		}
+		refVals = append(refVals, refVal)
 	}
 
-	if len(swComponents) == 0 {
+	if len(refVals) == 0 {
 		return nil, fmt.Errorf("no software components found")
 	}
 
-	return swComponents, nil
+	return refVals, nil
 }
 
-func makeSwAttrs(c PSAClassAttributes, s PSASwCompAttributes) (*structpb.Struct, error) {
-	swAttrs := map[string]interface{}{
-		"psa.impl-id":           c.ImplID,
-		"psa.signer-id":         s.SignerID,
-		"psa.measurement-value": s.MeasurementValue,
-		"psa.measurement-desc":  s.AlgID,
-	}
-
-	if c.Vendor != "" {
-		swAttrs["psa.hw-vendor"] = c.Vendor
-	}
+func ExtractMeas(obj MeasExtractor, m comid.Measurement, class PSAClassAttributes) (*proto.Endorsement, error) {
 
-	if c.Model != "" {
-		swAttrs["psa.hw-model"] = c.Model
+	if err := obj.FromMeasurement(m); err != nil {
+		return nil, err
 	}
 
-	if s.MeasurementType != "" {
-		swAttrs["psa.measurement-type"] = s.MeasurementType
+	refAttrs, err := obj.MakeRefAttrs(class)
+	if err != nil {
+		return &proto.Endorsement{}, fmt.Errorf("failed to create software component attributes: %w", err)
 	}
-
-	if s.Version != "" {
-		swAttrs["psa.version"] = s.Version
+	refVal := proto.Endorsement{
+		Scheme:     proto.AttestationFormat_PSA_IOT,
```

Comment:
as discussed f2f, we should have a separate scheme here

Reply:
Yes, I will evaluate this in the subsequent change!

Reply:
issue #90 Opened to track this!